NullReferenceException at launch in GUIMod constructor

**NOTE:** CKAN 1.25.1 runs just fine.  This is only an issue on 1.25.2.

<!--

Thanks for wanting to bring an issue to our attention!

Before you open a ticket, please do a quick search of the existing issues to see if it's already been brought up.

https://github.com/KSP-CKAN/CKAN/issues

-->

Background
----------

**CKAN Version:** 1.25.2


**KSP Version:** 1.4.5.2243


**Operating System:** macOS X 10.13.6 with Mono: Mono JIT compiler version 5.12.0.226 (2018-02/9824e260f56 Mon Apr 30 18:29:27 EDT 2018)


**Have you made any manual changes to your GameData folder (i.e., not via CKAN)?** NO


Problem
-------

**What steps did you take in CKAN?** Started it from command-line as I always do: mono ckan.exe


**What did you expect to happen?** CKAN to run and not crash.


**What happened instead?** CKAN Crashed with a stack trace (previous version, 1.25.1 runs just fine)


**Screenshots:**
<img width="664" alt="screen shot 2018-08-05 at 11 44 38 pm" src="https://user-images.githubusercontent.com/266627/43689897-8e882dd4-9909-11e8-862d-614d06d93426.png">



**CKAN error codes (if applicable):**

This happened under Linux Mint 18.3 as well.  I had updated mono to the 5.12 version when this happened, but I think the same thing also occurred when I tried the default 4.X version that comes with 18.3.

```
$ mono ckan.exe 
exception inside UnhandledException handler: Object reference not set to an instance of an object

[ERROR] FATAL UNHANDLED EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object
  at CKAN.GUIUser.DisplayMessage (System.String message, System.Object[] args) [0x00000] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.NullUser.RaiseMessage (System.String message, System.Object[] args) [0x00000] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.KSP.SetupCkanDirectories () [0x000dd] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.KSP..ctor (System.String gameDir, System.String name, CKAN.IUser user) [0x00038] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.KSPManager.LoadInstancesFromRegistry () [0x00060] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.KSPManager..ctor (CKAN.IUser user, CKAN.IWin32Registry win32_registry) [0x00028] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.Main..ctor (System.String[] cmdlineArgs, CKAN.KSPManager mgr, CKAN.GUIUser user, System.Boolean showConsole) [0x00039] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at (wrapper remoting-invoke-with-check) CKAN.Main..ctor(string[],CKAN.KSPManager,CKAN.GUIUser,bool)
  at CKAN.GUI.Main_ (System.String[] args, CKAN.KSPManager manager, System.Boolean showConsole) [0x0003b] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.CmdLine.MainClass.Gui (CKAN.KSPManager manager, CKAN.CmdLine.GuiOptions options, System.String[] args) [0x00008] in <352d74f237564d2e9ff313cc9ac1989e>:0 
  at CKAN.CmdLine.MainClass.Main (System.String[] args) [0x000a1] in <352d74f237564d2e9ff313cc9ac1989e>:0 
```
```$ mono --version
Mono JIT compiler version 5.12.0.301 (tarball Wed Jul 25 15:47:18 UTC 2018)
```

I was able to get CKAN functioning after applying the .deb package, and just running `ckan`

--- a/Core/Versioning/KspVersion.cs
+++ b/Core/Versioning/KspVersion.cs
@@ -2,7 +2,9 @@
 using System.Linq;
 using System.Text;
 using System.Text.RegularExpressions;
+using System.Collections.Generic;
 using Newtonsoft.Json;
+using CKAN.GameVersionProviders;
 
 namespace CKAN.Versioning
 {
@@ -228,6 +230,65 @@ namespace CKAN.Versioning
             return _string;
         }
 
+        private static Dictionary<string, KspVersion> VersionsMax = new Dictionary<string, KspVersion>();
+
+        /// <summary>
+        /// Generate version mapping table once for all instances to share
+        /// </summary>
+        static KspVersion()
+        {
+            // Should be sorted
+            List<KspVersion> versions = new KspBuildMap(new Win32Registry()).KnownVersions;
+            VersionsMax[""] = versions.Last();
+            foreach (var v in versions)
+            {
+                // Add or replace
+                VersionsMax[$"{v.Major}"          ] = v;
+                VersionsMax[$"{v.Major}.{v.Minor}"] = v;
+            }
+        }
+
+        /// <summary>
+        /// Get a string to represent a game version.
+        /// </summary>
+        /// <returns>
+        /// String representing max game version.
+        /// Partly clamped to real versions, partly rounded up to imaginary versions.
+        /// </returns>
+        public string ToYalovString()
+        {
+            KspVersion value;
+
+            if (!IsMajorDefined
+                // 2.0.0
+                || _major > VersionsMax[""].Major
+                // 1.99.99
+                || (_major == VersionsMax[""].Major && VersionsMax.TryGetValue($"{_major}", out value) && _minor >= UptoNines(value.Minor)))
+            {
+                return "any";
+            }
+            else if (IsMinorDefined
+                && VersionsMax.TryGetValue($"{_major}.{_minor}", out value)
+                && (!IsPatchDefined || _patch >= UptoNines(value.Patch)))
+            {
+                return $"{_major}.{_minor}.{UptoNines(value.Patch)}";
+            }
+            else
+            {
+                return ToString();
+            }
+        }
+
+        /// <returns>
+        ///   0 - 9   //  9  - 99    //  99 - 999
+        ///   1 - 9   //  10 - 99    // 100 - 999
+        ///   8 - 9   //  98 - 99
+        /// </returns>
+        private static int UptoNines(int num)
+        {
+            return (int)Math.Pow(10, Math.Floor(Math.Log10(num + 1)) + 1) - 1;
+        }
+
         /// <summary>
         /// Converts the value of the current <see cref="KspVersion"/> to its equivalent
         /// <see cref="KspVersionRange"/>.
@@ -239,7 +300,7 @@ namespace CKAN.Versioning
         /// </para>
         /// <para>
         /// For example, the version "1.0.0.0" would be equivalent to the range ["1.0.0.0", "1.0.0.0"], while the
-        /// version "1.0" would be equivalent to the range ["1.0.0.0", "1.1.0.0"). Where '[' and ']' represent 
+        /// version "1.0" would be equivalent to the range ["1.0.0.0", "1.1.0.0"). Where '[' and ']' represent
         /// inclusive bounds and '(' and ')' represent exclusive bounds.
         /// </para>
         /// </returns>
@@ -247,7 +308,7 @@ namespace CKAN.Versioning
         {
             KspVersionBound lower;
             KspVersionBound upper;
-        
+
             if (IsBuildDefined)
             {
                 lower = new KspVersionBound(this, inclusive: true);

--- a/GUI/GUIMod.cs
+++ b/GUI/GUIMod.cs
@@ -268,7 +268,7 @@ namespace CKAN
 
         private bool Equals(GUIMod other)
         {
-            return Equals(Name, other.Name);
+            return Equals(Identifier, other.Identifier);
         }
 
         public override bool Equals(object obj)
@@ -281,7 +281,7 @@ namespace CKAN
 
         public override int GetHashCode()
         {
-            return (Name != null ? Name.GetHashCode() : 0);
+            return Identifier?.GetHashCode() ?? 0;
         }
 
     }

--- a/GUI/GUIMod.cs
+++ b/GUI/GUIMod.cs
@@ -42,9 +42,6 @@ namespace CKAN
         public bool IsCKAN { get; private set; }
         public string Abbrevation { get; private set; }
 
-        private bool VersionMaxWasGenerated = false;
-        private Dictionary<string, KspVersion> VersionsMax;
-
         /// <summary>
         /// Return whether this mod is installable.
         /// Used for determining whether to show a checkbox in the leftmost column.
@@ -66,19 +63,6 @@ namespace CKAN
         public string Version
         {
             get { return IsInstalled ? InstalledVersion : LatestVersion; }
-        }
-
-        /// <returns>
-        ///   0 - 9   //  9  - 99    //  99 - 999    
-        ///   1 - 9   //  10 - 99    // 100 - 999
-        ///   8 - 9   //  98 - 99    // 
-        /// </returns>
-        private int UptoNines(int num)
-        {
-            //if (num == 0)
-            //    return 0;
-            //else
-                return (int)Math.Pow(10, Math.Floor(Math.Log10(num + 1)) + 1) - 1;
         }
 
         public GUIMod(CkanModule mod, IRegistryQuerier registry, KspVersionCriteria current_ksp_version, bool incompatible = false)
@@ -132,45 +116,8 @@ namespace CKAN
             // KSP.
             if (latest_available_for_any_ksp != null)
             {
-                if (!VersionMaxWasGenerated)
-                {
-                    VersionMaxWasGenerated = true;
-                    List<KspVersion> versions = new KspBuildMap(new Win32Registry()).KnownVersions;  // should be sorted
-
-                    VersionsMax = new Dictionary<string, KspVersion>();
-                    VersionsMax[""] = versions.Last();
-
-                    foreach (var v in versions)
-                    {
-                        VersionsMax[v.Major.ToString()] = v;        // add or replace
-                        VersionsMax[v.Major + "." + v.Minor] = v;   
-                    }
-                }
-
-                const int Undefined = -1;
-
-                KspVersion ksp_ver = registry.LatestCompatibleKSP(mod.identifier);
-                string ver = ksp_ver?.ToString();
-                int major = ksp_ver.Major, minor = ksp_ver.Minor, patch = ksp_ver.Patch;
-                KspVersion value;
-
-                if ( major == Undefined
-                    //|| (major >= UptoNines(VersionsMax[""].Major))       // 9.99.99
-                    || (major > VersionsMax[""].Major)                     // 2.0.0 
-                    || (major == VersionsMax[""].Major && VersionsMax.TryGetValue(major.ToString(), out value) && minor >= UptoNines(value.Minor))  // 1.99.99 ?
-                    )
-                    KSPCompatibility = "any";
-
-                else if (minor != Undefined
-                    && VersionsMax.TryGetValue(major + "." + minor, out value)
-                    && (patch == Undefined || patch >= UptoNines(value.Patch))
-                    )
-                    KSPCompatibility = major + "." + minor + "." + UptoNines(value.Patch);
-
-                else
-                    KSPCompatibility = ver;
-
-                // KSPCompatibility += " | " + major + "." + minor + "." + patch;   // for testing
+                KSPCompatibility = registry.LatestCompatibleKSP(mod.identifier)?.ToYalovString()
+                    ?? "Unknown";
 
                 // If the mod we have installed is *not* the mod we have installed, or we don't know
                 // what we have installed, indicate that an upgrade would be needed.
@@ -178,13 +125,11 @@ namespace CKAN
                 {
                     KSPCompatibilityLong = string.Format("{0} (using mod version {1})",
                         KSPCompatibility, latest_available_for_any_ksp.version);
-                    //    ver, latest_available_for_any_ksp.version);   //  true values in the right tab
 
                 }
                 else
                 {
                     KSPCompatibilityLong = KSPCompatibility;
-                    // KSPCompatibilityLong = ver;   //  true values in the right tab
                 }
             }
             else

